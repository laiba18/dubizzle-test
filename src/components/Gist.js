import Octicon from "react-octicon";
import PropTypes from 'prop-types';
import {
  DescriptionContainer,
  File,
  FilesContainer,
  GistHeader,
  GistWrapper,
  HeaderContainer,
  InfoContainer,
  InfoContainerText,
  OwnerImage,
  OwnerName,
  Timestamp,
  Timestamps
} from "./styles";
import { getDate } from "../utils/helper";

//  this function get files count
const getFileCount = (files) => Object.entries(files).length;

const Gist = ({ gist }) => {
  //  extract data from gist object and display below.
  const {
    owner: { avatar_url, login, starred_url, html_url } = {},
    files = [],
    created_at,
    updated_at,
    description,
    forks_url,
    comments_url
  } = gist || {};

  // date list array of objects used to render the multiple data.
  const dateList = [
    {
      label: `Created at: ${getDate(created_at)}`
    },
    {
      label: `Last updated: ${getDate(updated_at)}`
    }
  ];

  // information list array of objects used to render the multiple data and link values also.
  const infoDataList = [
    {
      label: `${getFileCount(files)}
      ${getFileCount(files) > 1 ? "Files" : "File"}`,
      name: "code",
    },
    {
      label: "Forks",
      name: "repo-forked",
      link: forks_url
    },
    {
      label: "Comments",
      name: "comment",
      link: comments_url
    },
    {
      label: "Stars",
      name: "star",
      link: starred_url
    }
  ]

  return (
    <GistWrapper>
      <GistHeader>
        <HeaderContainer>
          {/* owner information display */}
          <OwnerImage src={avatar_url} />
          <a href={html_url} target="_blank" rel="noreferrer">
            <OwnerName>{login}</OwnerName>
          </a>
        </HeaderContainer>
        <HeaderContainer>
          {/* info display items */}
          {
            infoDataList.map((infoItem, index) => (
              <InfoContainer key={index}>
                <Octicon name={infoItem.name} />
                <a href={infoItem.link} target="_blank" rel="noreferrer">
                  <InfoContainerText>
                    {infoItem.label}
                  </InfoContainerText>
                </a>

              </InfoContainer>
            ))
          }
        </HeaderContainer>
      </GistHeader>
      <Timestamps>
        {/* created and updated display */}
        {
          dateList.map((dateItem, index) => (
            <Timestamp key={index}>
              {dateItem.label}
            </Timestamp>
          ))
        }
      </Timestamps>
      <DescriptionContainer>{description}</DescriptionContainer>
      <FilesContainer>
        {/* files name display */}
        {Object.keys(files).map((fileItem) => (
          <File>
            <Octicon name="file-text" />
            {fileItem}
          </File>
        ))}
      </FilesContainer>
    </GistWrapper>
  );
};

// Declare that a prop is a specific JS type. By default, these are required
Gist.propTypes = {
  gist: PropTypes.object.isRequired
}

export default Gist;
