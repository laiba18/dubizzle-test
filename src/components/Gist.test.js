import { render } from "@testing-library/react";
import Gist from "./Gist";

const gist = {
  owner: {
    login
      : "maximymym"
  }
}

describe("Gist test cases", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<Gist gist={gist} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
