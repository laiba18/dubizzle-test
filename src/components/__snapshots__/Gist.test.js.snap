// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gist test cases should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dicizt ipiXZh"
  >
    <div
      class="sc-iVCKna bjJGBQ"
    >
      <div
        class="sc-dcrmVg jmCkWd"
      >
        <img
          class="sc-gxYJeL hIzBgj"
        />
        <a
          rel="noreferrer"
          target="_blank"
        >
          <p
            class="sc-kPTPQs kuuQoX"
          />
        </a>
      </div>
      <div
        class="sc-dcrmVg jmCkWd"
      >
        <div
          class="sc-ispOId fcUjBH"
        >
          <span
            class="octicon octicon-code"
          />
          <a
            rel="noreferrer"
            target="_blank"
          >
            <div
              class="sc-kcuKUB iwGKyj"
            >
              0
      File
            </div>
          </a>
        </div>
        <div
          class="sc-ispOId fcUjBH"
        >
          <span
            class="octicon octicon-repo-forked"
          />
          <a
            rel="noreferrer"
            target="_blank"
          >
            <div
              class="sc-kcuKUB iwGKyj"
            >
              Forks
            </div>
          </a>
        </div>
        <div
          class="sc-ispOId fcUjBH"
        >
          <span
            class="octicon octicon-comment"
          />
          <a
            rel="noreferrer"
            target="_blank"
          >
            <div
              class="sc-kcuKUB iwGKyj"
            >
              Comments
            </div>
          </a>
        </div>
        <div
          class="sc-ispOId fcUjBH"
        >
          <span
            class="octicon octicon-star"
          />
          <a
            rel="noreferrer"
            target="_blank"
          >
            <div
              class="sc-kcuKUB iwGKyj"
            >
              Stars
            </div>
          </a>
        </div>
      </div>
    </div>
    <div
      class="sc-kuWgmH la-DjjL"
    >
      <p
        class="sc-duSInm fcsswO"
      >
        Created at: Invalid Date
      </p>
      <p
        class="sc-duSInm fcsswO"
      >
        Last updated: Invalid Date
      </p>
    </div>
    <div
      class="sc-jrrXlR kAKLgO"
    />
    <div
      class="sc-crozmw jxZFmi"
    />
  </div>
</DocumentFragment>
`;
